### ISSUE TITLE
No LDAP User <-> Group Mapping

### ISSUE THREAD

**Issue by sbertolin-iks** [Graylog2/graylog2-server#2461]
### Problem description

I've configured LDAP external autentication and it works fine, but the correlation between ldap users and ldap groups is not fetched fine. I see that in groups configuration there is no way to configure attribute to map member groups. In my case I have to use "member" attribute of ldap group object (see the screenshots).
Thank you.
Best regards.

### Steps to reproduce the problem
1. ...
### Environment
- Graylog Version:
- Elasticsearch Version:
- MongoDB Version:
- Operating System:
- Browser version:

**Comment by kroepke** [comment]
Graylog currently tries to autodetect which attribute to use based on the objectClass used for groups:
* `groupOfUniqueNames` uses `uniqueMember`
* `groupOfNames` and `group` use `member`
* `posixGroup` uses `memberUid`

Based on what I see from the screenshot it should work in your case.

Graylog then takes the authenticated user's `dn` and tries to find it in the member list by comparing the entries.

To better debug this you can set the log level for `org.graylog2.security.ldap.LdapConnector` to `trace` in an external log4j2 configuration as described in the documentation: http://docs.graylog.org/en/2.0/pages/installation/manual_setup.html#supplying-external-logging-configuration

Then attempt another login, this will write quite a bit of detail to the server log file, including lines like `DN {} == {} member?`
This should give you an indication whether this is a bug or misconfiguration.

If you have additional information please update this bug report.

Thanks!

**Comment by sbertolin-iks** [comment]
Hi,
Ive configured logger as you suggested and in logs seems that my ldap User e3sberto is mapped fine with "wiki_user" ldap group. But I don't see correlation with "AGW Developer" role grants, that is correlated to that group. I attach log file and ui screenshots.
Thank you.
Best regards.

2016-07-06 18:16:16,553 TRACE: org.graylog2.security.ldap.LdapConnector - Group Entry:   Entry
      dn: cn=wiki_user,cl=Gruppi Websphere,cl=Account,og=Gruppo Generali
      objectClass: Top
      objectClass: groupOfNames
      uniqueMember: uid=e3etrevi@GGS,cl=GGS,cl=Account,og=Gruppo Generali
      uniqueMember: uid=e3dpuzzo@GGS,cl=GGS,cl=Account,og=Gruppo Generali
      uniqueMember: uid=bmc0001@GGS,cl=GGS,cl=Account,og=Gruppo Generali
      uniqueMember: uid=e3aquagg@GGS,cl=GGS,cl=Account,og=Gruppo Generali
      uniqueMember: uid=pasquat1@GGS,cl=GGS,cl=Account,og=Gruppo Generali
      uniqueMember: uid=caraffa@GGS,cl=GGS,cl=Account,og=Gruppo Generali
      uniqueMember: uid=e3sfiori@GGS,cl=GGS,cl=Account,og=Gruppo Generali
      uniqueMember: uid=arusso@GGS,cl=GGS,cl=Account,og=Gruppo Generali
      uniqueMember: uid=e3gmastr@GGS,cl=GGS,cl=Account,og=Gruppo Generali
      uniqueMember: uid=mbin@GGS,cl=GGS,cl=Account,og=Gruppo Generali
      uniqueMember: uid=tonetto@GGS,cl=GGS,cl=Account,og=Gruppo Generali
      uniqueMember: uid=greco10@GGS,cl=GGS,cl=Account,og=Gruppo Generali
      uniqueMember: uid=bonato@GGS,cl=GGS,cl=Account,og=Gruppo Generali
      uniqueMember: uid=ferla@GGS,cl=GGS,cl=Account,og=Gruppo Generali
      uniqueMember: uid=simiona3@GGS,cl=GGS,cl=Account,og=Gruppo Generali
      uniqueMember: uid=bonazza@GGS,cl=GGS,cl=Account,og=Gruppo Generali
      uniqueMember: uid=frison@GGS,cl=GGS,cl=Account,og=Gruppo Generali
      uniqueMember: uid=filippi@GGS,cl=GGS,cl=Account,og=Gruppo Generali
      uniqueMember: uid=e3aalbe1@GGS,cl=GGS,cl=Account,og=Gruppo Generali
      uniqueMember: uid=e3mrizzi@GGS,cl=GGS,cl=Account,og=Gruppo Generali
      uniqueMember: uid=e3aross4@GGS,cl=GGS,cl=Account,og=Gruppo Generali
      uniqueMember: uid=e3acresp@GGS,cl=GGS,cl=Account,og=Gruppo Generali
      uniqueMember: uid=e3dabbal@GGS,cl=GGS,cl=Account,og=Gruppo Generali
      uniqueMember: uid=e3fcerut@GGS,cl=GGS,cl=Account,og=Gruppo Generali
      uniqueMember: uid=e3gvici1@GGS,cl=GGS,cl=Account,og=Gruppo Generali
      uniqueMember: uid=e3fnalet@GGS,cl=GGS,cl=Account,og=Gruppo Generali
      uniqueMember: uid=e3lbassa@GGS,cl=GGS,cl=Account,og=Gruppo Generali
      uniqueMember: uid=grigolet@GGS,cl=GGS,cl=Account,og=Gruppo Generali
      uniqueMember: uid=e3branzi@GGS,cl=GGS,cl=Account,og=Gruppo Generali
      uniqueMember: uid=e3mfinoc@GGS,cl=GGS,cl=Account,og=Gruppo Generali
      uniqueMember: uid=e3ldicre@GGS,cl=GGS,cl=Account,og=Gruppo Generali
      uniqueMember: uid=baldo@GGS,cl=GGS,cl=Account,og=Gruppo Generali
      uniqueMember: uid=e3sberto@GGS,cl=GGS,cl=Account,og=Gruppo Generali
      cn: wiki_user
![graylog-screenshot2](https://cloud.githubusercontent.com/assets/20317977/16626065/9503224a-43a7-11e6-8627-3e6625cf4fc1.PNG)
![graylog-screenshot](https://cloud.githubusercontent.com/assets/20317977/16626064/9502bfc6-43a7-11e6-8478-7569c5611c33.PNG)
![graylog-screenshot3](https://cloud.githubusercontent.com/assets/20317977/16626066/950364c6-43a7-11e6-9111-05fe56e95f1c.PNG)
[graylog-server.zip](https://github.com/Graylog2/graylog2-server/files/350383/graylog-server.zip)

**Comment by kroepke** [comment]
Weirdly the group entries do _not_ seem to have a `member` attribute but rather `uniqueMembers`.
Graylog currently doesn't allow setting the attribute to search for, we simply follow the core schema as I describe above: http://www.zytrax.com/books/ldap/ape/core-schema.html#groupofnames

So basically everything works but we are expecting a different attribute. What kind of LDAP server is this? Active Directory?
I've never seen `groupOfNames` return `uniqueMember` and from the screenshots it looks like Apache Directory sees it as `member`, too.

**Comment by sbertolin-iks** [comment]
Hi,
here is LDAP product name and version: **Novell eDirectory** (https://www.novell.com/documentation/edir873/?page=/documentation/edir873/edir873/data/h0000007.html ).
Product version: 
Binary Version: 20807.13
Root Most Entry Depth: 0
Product Version: **eDirectory for Linux x86_64 v8.8 SP8** [DS]

Here is detailed list of installed packages:
ldapmg07:~ # rpm -qa|grep novell
novell-NDSbase-8.8.8.6-0
novell-NDSserv-8.8.8.6-0
novell-NOVLsnmp-8.8.8.6-0
novell-npkiapi-8.8.8.6-20150915_1117
novell-edirectory-expat-8.8.8-1
novell-edirectory-xdaslog-conf-8.8.8.3-7
novell-NLDAPbase-8.8.8.6-0
novell-npkit-8.8.8.6-20150915_1116
novell-NOVLsubag-8.8.8.6-0
novell-NOVLlmgnt-8.8.8.6-0
novell-NOVLldif2dib-8.8.8.6-0
novell-edirectory-xdasinstrument-8.8.8.6-0
novell-kerberos-base-1.5-54.1
novell-NDSmasv-8.8.8.6-0
novell-NDScommon-8.8.8.6-0
novell-ntls-8.8.8.5-150714
novell-nmas-8.8.8.6-20150915
novell-NOVLembox-8.8.8.6-4
novell-edirectory-jclnt-8.8.8.6-0
novell-NOVLice-8.8.8.6-0
novell-kerberos-ldap-extensions-1.5-54.1
novell-pkiserver-8.8.8.6-20150915_1118
novell-NOVLxis-8.8.8.6-0
novell-edirectory-xdaslog-8.8.8.3-7
novell-DXMLfanplat-4.0-0
novell-NLDAPsdk-8.8.8.6-0
novell-NDSrepair-8.8.8.6-5
novell-NDSimon-8.8.8.6-0
novell-ncpenc-1.2.2.0-0

**Comment by sbertolin-iks** [comment]
Hi,
here is LDAP product name and version: Novell eDirectory (https://www.novell.com/documentation/edir873/?page=/documentation/edir873/edir873/data/h0000007.html ).
Product version:
Binary Version: 20807.13
Root Most Entry Depth: 0
Product Version: eDirectory for Linux x86_64 v8.8 SP8 [DS]

Here is detailed list of installed packages:
ldapmg07:~ # rpm -qa|grep novell
novell-NDSbase-8.8.8.6-0
novell-NDSserv-8.8.8.6-0
novell-NOVLsnmp-8.8.8.6-0
novell-npkiapi-8.8.8.6-20150915_1117
novell-edirectory-expat-8.8.8-1
novell-edirectory-xdaslog-conf-8.8.8.3-7
novell-NLDAPbase-8.8.8.6-0
novell-npkit-8.8.8.6-20150915_1116
novell-NOVLsubag-8.8.8.6-0
novell-NOVLlmgnt-8.8.8.6-0
novell-NOVLldif2dib-8.8.8.6-0
novell-edirectory-xdasinstrument-8.8.8.6-0
novell-kerberos-base-1.5-54.1
novell-NDSmasv-8.8.8.6-0
novell-NDScommon-8.8.8.6-0
novell-ntls-8.8.8.5-150714
novell-nmas-8.8.8.6-20150915
novell-NOVLembox-8.8.8.6-4
novell-edirectory-jclnt-8.8.8.6-0
novell-NOVLice-8.8.8.6-0
novell-kerberos-ldap-extensions-1.5-54.1
novell-pkiserver-8.8.8.6-20150915_1118
novell-NOVLxis-8.8.8.6-0
novell-edirectory-xdaslog-8.8.8.3-7
novell-DXMLfanplat-4.0-0
novell-NLDAPsdk-8.8.8.6-0
novell-NDSrepair-8.8.8.6-5
novell-NDSimon-8.8.8.6-0
novell-ncpenc-1.2.2.0-0



Best regards



Stefano Bertolin
Senior Consultant
Mobile: +39 3358324811

IKS srl
Tel: +39 049 8701010|Fax: +39 049 8701014
Padova: C.so Stati Uniti, 14 bis - 35127
Milano: Sesto S. Giovanni, Piazza Martiri di Via Fani, 19 - 20099
Trento: Rovereto, C.so Rosmini, 8 - 38068
Web: www.iks.it<http://www.iks.it/> |www.kimaps.it<http://www.kimaps.it/> | www.xtn-lab.com<http://www.xtn-lab.com/> | www.kirey.it<http://www.kirey.it/>
Follow us: Facebook<https://www.facebook.com/IKSGROUP/> | Twitter<https://twitter.com/IKSgroup>

The information in this e-mail message, and any file transmitted with it, are confidential and privileged.
P please consider the environment before printing this e-mail

Da: Kay Roepke [mailto:notifications@github.com]
Inviato: giovedì 7 luglio 2016 08:12
A: Graylog2/graylog2-server
Cc: Stefano Bertolin; Author
Oggetto: Re: [Graylog2/graylog2-server] No LDAP User <-> Group Mapping (#2461)


Weirdly the group entries do not seem to have a member attribute but rather uniqueMembers.
Graylog currently doesn't allow setting the attribute to search for, we simply follow the core schema as I describe above: http://www.zytrax.com/books/ldap/ape/core-schema.html#groupofnames

So basically everything works but we are expecting a different attribute. What kind of LDAP server is this? Active Directory?
I've never seen groupOfNames return uniqueMember and from the screenshots it looks like Apache Directory sees it as member, too.

**Comment by sbertolin-iks** [comment]
Hi,
any news?
Thank you
Best regards

**Comment by johnseekins** [comment]
So I have a similar problem with FreeIPA.

It provides `posixGroup` as the roleObjectClass, but `member` as the roleMemberAttribute.

Any reason this couldn't just be a configurable option?

**Comment by resyst-it** [comment]
I have a the same problem with freeipa.

I can specify posixgroup or groupings, in the logs I always have a search on the member attribute.

**Comment by Kaydub00** [comment]
I have this same issue in FreeIPA. I can retrieve groups and map them to Graylog roles but when someone logs in I get the message [LdapConnector] Unable to auto-detect the LDAP group object class, assuming 'member' is the correct attribute.

Right now I've removed any of the ldap member configuration and am relying on configuring the groups in Graylog.

**Comment by resyst-it** [comment]
I found the solution to my problem.
Watch out for groups. In freeip it is cn=compat.
Exemple : cn=groups,cn=compat,dc=exemple,dc=com

You are not cn=compat with users, only with groups.

**Comment by fuero** [comment]
I'm experiencing this bug with Novell eDirectory as well. As far as I can tell, the trouble is that the schema Novell ships with the product (i.e. their interpretation of the RFC 2256) considers `uniqueMember` and `member` equivalent.

Explicitly searching for one and the other attribute will list the values with the one or the other attribute, respectively. Combining the 2 in one query will yield `uniqueMember`, thus confusing Graylog's detection code.

This is precisely why code dealing with LDAP should not assume attributes or objectclasses for particular implementations. Detection and profiles could be provided as hints for the admin setting it up, but the vital settings like the attributes responsible for group mapping should be configurable.

So my proposal is this:

Include config settings for the group-to-user mapping attribute and for the user-to-group mapping attribute and change the LDAP code to adhere to them. If this is possible already, nothing in the docs I found hints at it.

**Comment by bernd** [comment]
This should no longer be an issue in 4.0 because we are using a different method to lookup groups for a user. Closing this issue.
